src/main/java/com/direwolf20/buildinggadgets/common/building/Region.java

```diff
@@ -0,0 +1,345 @@
+package com.direwolf20.buildinggadgets.common.building;
+
+import com.direwolf20.buildinggadgets.common.tools.GadgetUtils;
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.PeekingIterator;
+import net.minecraft.util.math.AxisAlignedBB;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Vec3i;
+
+import java.util.Objects;
+import java.util.Spliterator;
+
+/**
+ * Represents a region in the world with a finite and nonzero size.
```

Comment:
```suggestion
 * Represents a region in the world of finite size.
```
As you are not enforcing 0 size, that does not belong here. (Also please don't enforce it, as that would differ from me allowing the Existence of the Empty Template as a replacement for null)

Reply:
There's no way you can make a 0 size region with the current implementation. When the two vertexes equal to each other, the size will be 1 (which would make sense as the vertexes has volume).
One possibility would be create a `public static final Region EMPTY = new Region(...) {...}` where I override the methods to create the correct behavior.

Reply:
Sorry I completly forgot about adding 1 to each vertex lenght... Which makes totally sense...
Just write `of` then...

Reply:
Here's the problem: `Region` is final, which means no anonymous classes can be created. I think the only way to make it work is by having `Region` an interface, and a `DefaultRegion` implements the interface `Region` which can be overridden.

Reply:
Additionally, since we already have template as an interface, we can easily create an empty template reference that overrides the methods to work like an actual empty template.

Reply:
There is no need to make this so complicated - please leave region as is...
A one block region is close enough to an empty region - I'll just make the Template like that...